src/coreclr/jit/lower.cpp

```diff
@@ -3777,12 +3777,41 @@ void Lowering::LowerRetStruct(GenTreeUnOp* ret)
 
         case GT_BLK:
         case GT_OBJ:
-            retVal->ChangeOper(GT_IND);
-            FALLTHROUGH;
         case GT_IND:
+        {
+            // Spill to a local if sizes don't match so we can avoid the "load more than requested"
+            // problem, e.g. struct size is 5 and we emit "ldr x0, [x1]"
+            unsigned             realSize  = retVal->AsIndir()->Size();
+            CORINFO_CLASS_HANDLE structCls = comp->info.compMethodInfo->args.retTypeClass;
+            if (realSize == 0)
+            {
+                // We have an IND<struct> node, assuming it's used in a return statement, take target size
+                // from the current method's signature (it's only used to decide whether we need to spill it to
+                // a local or not, so the actual size won't be used).
+                // TODO-ADDR: delete once "IND<struct>" nodes are no more
+                realSize = comp->info.compCompHnd->getClassSize(structCls);
+            }
+
+            if (genTypeSize(nativeReturnType) > realSize)
+            {
+                LIR::Use retValUse(BlockRange(), &ret->gtOp1, ret);
+                unsigned tmpNum = BAD_VAR_NUM;
+                if (structCls != NO_CLASS_HANDLE)
+                {
+                    tmpNum               = comp->lvaGrabTemp(true DEBUGARG("mis-sized struct return"));
+                    comp->genReturnLocal = tmpNum;
+                    comp->lvaSetStruct(tmpNum, structCls, true);
+                }
```

Comment:
```suggestion
                LIR::Use retValUse(BlockRange(), &ret->gtOp1, ret);
                unsigned tmpNum = comp->lvaGrabTemp(true DEBUGARG("mis-sized struct return"));
                comp->lvaSetStruct(tmpNum, structCls, false);
```

1) `genReturnLocal` has a very specific purpose, no reason to set it here.
2) No unsafe value class checks needed (in principle, in practice it doesn't matter) since the address of this local won't escape.

Reply:
> genReturnLocal has a very specific purpose, no reason to set it here.

Tests assert after this patch: Assertion failed 'tmp == genReturnLocal'

Reply:
I reverted some suggestions if you don't mind because we hit the path for non-structs as well so now it's cleaner that we have a special case for that rare case with IND<struct>.

Reply:
We still need to handle the case with handle-less `BLK` nodes though.

Reproduction:
```cs
struct Sx3
{
    public byte A, B, C;
}

[MethodImpl(MethodImplOptions.NoInlining)]
private static Sx3 Problem(StructWithIndex* s)
{
    Sx3 a;
    Unsafe.CopyBlock(&a, s, 3);
    return a;
}
```

Reply:
thanks, reverted the revert. @SingleAccretion does it look good to you? I assume it's just your change written with my hands 😄